quic/s2n-quic-core/src/packet/interceptor.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{event::api::Subject, packet::number::PacketNumber, time::Timestamp};
+use s2n_codec::{DecoderBufferMut, EncoderBuffer};
+
+#[derive(Debug)]
+pub struct Packet {
+    pub number: PacketNumber,
+    pub timestamp: Timestamp,
+}
+
+/// Trait which enables an application to intercept packets that are transmitted and received
+pub trait Interceptor: 'static + Send {
+    #[inline(always)]
+    fn intercept_rx_payload<'a>(
+        &mut self,
+        subject: Subject,
+        packet: Packet,
+        payload: DecoderBufferMut<'a>,
+    ) -> DecoderBufferMut<'a> {
+        let _ = subject;
+        let _ = packet;
+        payload
+    }
+
+    #[inline(always)]
+    fn intercept_tx_payload<'a>(
+        &mut self,
+        subject: Subject,
+        packet: Packet,
+        payload: &mut EncoderBuffer<'a>,
+    ) {
+        let _ = subject;
+        let _ = packet;
+        let _ = payload;
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct Disabled(());
```

Comment:
could this just be `pub struct Disabled;`?

Reply:
It could be. The nice thing about making it a tuple struct with a private field is it's still ZST but can only be constructed with `Default`. While unlikely, this allows us to add fields if we even need it.